File: 155-min-stack/155-min-stack.java
class MinStack {
    Stack<Long> st;
    long min;
    
    public MinStack() {
        st=new Stack<>();
        min=Long.MAX_VALUE;
    }
    
    public void push(int x) {
        if(st.size()==0){
            min=(long)x;
            st.push((long)x);
            return;
        }    
        
        if(x<min){
            long to_push=((long)(x)-min)+(long)(x); 
            
            st.push(to_push);
            min=((long)(x));
        } else {
            st.push((long)(x));
        }
    }
    
    public void pop() {
        if(st.peek()<min){ // we are popping encrypted value
            long tp=st.pop();
            
            long x=min;
            long old_min=2*x-tp;
            
            min=old_min;
        } else {
            st.pop();
        }
    }
    
    public int top() {
         long tr;
        if(st.peek()<min){ 
            tr=min;
        } else {
            tr=st.peek();
        }
        
        return (int)tr;
    }
    
    public int getMin() {
        return (int)min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 155-min-stack/155-min-stack.cpp
class MinStack {
public:
    stack<long> st;
    long mine=0;
    MinStack() {
        
    }
    
    void push(int val) {
        if(st.size()==0){
            mine=(long)val;
            st.push((long)val);
        } else if(val<mine){
            long tp=((long)val-mine)+(long)val; // tp<val;
            st.push(tp);
            
            mine=(long)val;
        } else {
            st.push((long)val);
        }
    }
    
    void pop() {
        if(st.top()<mine){
            long val=mine;
            long tp=st.top();
            
            long old_min= val - tp + val;
            
            mine=old_min;
        }
        st.pop();
    }
    
    int top() {
        if(st.top()<mine){
            return (int)(mine);
        } else {
            return (int)st.top();
        }
    }
    
    int getMin() {
        return (int)mine;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */